opamp/observiq/agent_config_manager.go

```diff
@@ -0,0 +1,167 @@
+func (a *AgentConfigManager) ComposeEffectiveConfig() (*protobufs.EffectiveConfig, error) {
+	contentMap := make(map[string]*protobufs.AgentConfigFile, len(a.configMap))
+
+	// Used to track config names to alphabetize later in hash compution
+	configNames := make([]string, 0, len(a.configMap))
+
+	for configName, managedConfig := range a.configMap {
+		// Read in config file
+		cleanPath := filepath.Clean(managedConfig.ConfigPath)
+		configContents, err := os.ReadFile(cleanPath)
+		if err != nil {
+			return nil, fmt.Errorf("error reading config file %s: %w", configName, err)
+		}
+
+		// Add to contentMap
+		contentMap[configName] = &protobufs.AgentConfigFile{
+			Body:        configContents,
+			ContentType: opamp.DetermineContentType(managedConfig.ConfigPath),
+		}
+
+		configNames = append(configNames, configName)
+	}
+
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: contentMap,
+		},
+	}, nil
+}
+
+// ApplyConfigChanges compares the remoteConfig to the existing and applies changes
+func (a *AgentConfigManager) ApplyConfigChanges(remoteConfig *protobufs.AgentRemoteConfig) (effectiveConfig *protobufs.EffectiveConfig, changed bool, err error) {
+	effectiveConfig, err = a.ComposeEffectiveConfig()
+	if err != nil {
+		return nil, false, err
+	}
+
+	currentConfigMap := effectiveConfig.GetConfigMap().GetConfigMap()
+	remoteConfigMap := remoteConfig.GetConfig().GetConfigMap()
+
+	// No remote config Map
+	if remoteConfigMap == nil {
+		return
+	}
+
+	// loop through all remote configs and compare then with existing configs
+	for configName, remoteContents := range remoteConfigMap {
+		// For security check the log file we want is acceptable
+		if _, ok := acceptableConfigs[configName]; !ok {
+			a.logger.Info("Not support config received skipping", zap.String("config", configName))
+			continue
+		}
+
+		currentContents, ok := currentConfigMap[configName]
+		if !ok {
+			a.logger.Info("Untracked config found", zap.String("config", configName))
+			// We don't current track this config file we should add it
+			changed = true
+
+			// Write out the file
+			if err := os.WriteFile(configName, remoteContents.GetBody(), 0600); err != nil {
+				return nil, false, fmt.Errorf("failed to write new config file %s: %w", configName, err)
+			}
+
+			// Track new config
+			a.AddConfig(configName, &opamp.ManagedConfig{
+				ConfigPath: filepath.Join(".", configName),
+				Reload:     opamp.NoopReloadFunc,
+			})
+			continue
+		}
+
+		// Check to see if file contents are the same
+		if bytes.Equal(currentContents.GetBody(), remoteContents.GetBody()) {
+			continue
+		}
+
+		// Update the config file
+		managedConfig := a.configMap[configName]
+		newContents := remoteContents.GetBody()
+
+		a.logger.Info("Applying changes to config file", zap.String("config", configName))
+		configChanged, err := managedConfig.Reload(newContents)
```

Comment:
would there ever be a case where there was new config data, but this returned NO config change and NO error? in this case we'd just quietly move on and the end result of this function could be no change no error and not even a log message explaining why the heck the config didn't change when there was in fact a remote change.

does that make sense?

Reply:
I'm not sure I follow. If there's new data we trigger the reload function. If a reload fails we submit and error. We're trusting the reload function to do its job. If it decides there's no changes then there's nothing to report. The spec is pretty open ended on this. It even calls out the agent doesn't have to take any configuration changes from the platform.

I get what you're saying but not sure I see a case where that could happen.

Reply:
@cpheps i'm just saying that this method doesn't have control over how the managedConfig.Reload method is implemented. but i guess if we're saying this is a false condition that should never be implemented that way, then sure nothing necessarily needs to be done here